File: 12_Bombs_Away/javascript/bombsaway.js
// BOMBS AWAY
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    while (1) {
        print("YOU ARE A PILOT IN A WORLD WAR II BOMBER.\n");
        while (1) {
            print("WHAT SIDE -- ITALY(1), ALLIES(2), JAPAN(3), GERMANY(4)");
            a = parseInt(await input());
            if (a < 1 || a > 4)
                print("TRY AGAIN...\n");
            else
                break;
        }
        if (a == 1) {
            while (1) {
                print("YOUR TARGET -- ALBANIA(1), GREECE(2), NORTH AFRICA(3)");
                b = parseInt(await input());
                if (b < 1 || b > 3)
                    print("TRY AGAIN...\n");
                else
                    break;
            }
            print("\n");
            if (b == 1) {
                print("SHOULD BE EASY -- YOU'RE FLYING A NAZI-MADE PLANE.\n");
            } else if (b == 2) {
                print("BE CAREFUL!!!\n");
            } else {
                print("YOU'RE GOING FOR THE OIL, EH?\n");
            }
        } else if (a == 2) {
            while (1) {
                print("AIRCRAFT -- LIBERATOR(1), B-29(2), B-17(3), LANCASTER(4)");
                g = parseInt(await input());
                if (g < 1 || g > 4)
                    print("TRY AGAIN...\n");
                else
                    break;
            }
            print("\n");
            if (g == 1) {
                print("YOU'VE GOT 2 TONS OF BOMBS FLYING FOR PLOESTI.\n");
            } else if (g == 2) {
                print("YOU'RE DUMPING THE A-BOMB ON HIROSHIMA.\n");
            } else if (g == 3) {
                print("YOU'RE CHASING THE BISMARK IN THE NORTH SEA.\n");
            } else {
                print("YOU'RE BUSTING A GERMAN HEAVY WATER PLANT IN THE RUHR.\n");
            }
        } else if (a == 3) {
            print("YOU'RE FLYING A KAMIKAZE MISSION OVER THE USS LEXINGTON.\n");
            print("YOUR FIRST KAMIKAZE MISSION(Y OR N)");
            str = await input();
            if (str == "N") {
                s = 0;
            } else {
                s = 1;
                print("\n");
            }
        } else {
            while (1) {
                print("A NAZI, EH?  OH WELL.  ARE YOU GOING FOR RUSSIA(1),\n");
                print("ENGLAND(2), OR FRANCE(3)");
                m = parseInt(await input());
                if (m < 1 || m > 3)
                    print("TRY AGAIN...\n");
                else
                    break;
            }
            print("\n");
            if (m == 1) {
                print("YOU'RE NEARING STALINGRAD.\n");
            } else if (m == 2) {
                print("NEARING LONDON.  BE CAREFUL, THEY'VE GOT RADAR.\n");
            } else if (m == 3) {
                print("NEARING VERSAILLES.  DUCK SOUP.  THEY'RE NEARLY DEFENSELESS.\n");
            }
        }
        if (a != 3) {
            print("\n");
            while (1) {
                print("HOW MANY MISSIONS HAVE YOU FLOWN");
                d = parseInt(await input());
                if (d < 160)
                    break;
                print("MISSIONS, NOT MILES...\n");
                print("150 MISSIONS IS HIGH EVEN FOR OLD-TIMERS.\n");
                print("NOW THEN, ");
            }
            print("\n");
            if (d >= 100) {
                print("THAT'S PUSHING THE ODDS!\n");
            } else if (d < 25) {
                print("FRESH OUT OF TRAINING, EH?\n");
            }
            print("\n");
            if (d >= 160 * Math.random())
                hit = true;
            else
                hit = false;
        } else {
            if (s == 0) {
                hit = false;
            } else if (Math.random() > 0.65) {
                hit = true;
            } else {
                hit = false;
                s = 100;
            }
        }
        if (hit) {
            print("DIRECT HIT!!!! " + Math.floor(100 * Math.random()) + " KILLED.\n");
            print("MISSION SUCCESSFUL.\n");
        } else {
            t = 0;
            if (a != 3) {
                print("MISSED TARGET BY " + Math.floor(2 + 30 * Math.random()) + " MILES!\n");
                print("NOW YOU'RE REALLY IN FOR IT !!\n");
                print("\n");
                while (1) {
                    print("DOES THE ENEMY HAVE GUNS(1), MISSILE(2), OR BOTH(3)");
                    r = parseInt(await input());
                    if (r < 1 || r > 3)
                        print("TRY AGAIN...\n");
                    else
                        break;
                }
                print("\n");
                if (r != 2) {
                    print("WHAT'S THE PERCENT HIT RATE OF ENEMY GUNNERS (10 TO 50)");
                    s = parseInt(await input());
                    if (s < 10)
                        print("YOU LIE, BUT YOU'LL PAY...\n");
                    print("\n");
                }
                print("\n");
                if (r > 1)
                    t = 35;
            }
            if (s + t <= 100 * Math.random()) {
                print("YOU MADE IT THROUGH TREMENDOUS FLAK!!\n");
            } else {
                print("* * * * BOOM * * * *\n");
                print("YOU HAVE BEEN SHOT DOWN.....\n");
                print("DEARLY BELOVED, WE ARE GATHERED HERE TODAY TO PAY OUR\n");
                print("LAST TRIBUTE...\n");
            }
        }
        print("\n");
        print("\n");
        print("\n");
        print("ANOTHER MISSION (Y OR N)");
        str = await input();
        if (str != "Y")
            break;
    }
    print("CHICKEN !!!\n");
    print("\n");
}

main();

File: 12_Bombs_Away/python/bombs_away.py
"""
Bombs away

Ported from BASIC to Python3 by Bernard Cooke (bernardcooke53)
Tested with Python 3.8.10, formatted with Black and type checked with mypy.
"""
import random
from typing import Iterable


def _stdin_choice(*, prompt: str, choices: Iterable[str]) -> str:
    ret = input(prompt)
    while ret not in choices:
        print("TRY AGAIN...")
        ret = input(prompt)
    return ret


def player_survived() -> None:
    print("YOU MADE IT THROUGH TREMENDOUS FLAK!!")


def player_death() -> None:
    print("* * * * BOOM * * * *")
    print("YOU HAVE BEEN SHOT DOWN.....")
    print("DEARLY BELOVED, WE ARE GATHERED HERE TODAY TO PAY OUR")
    print("LAST TRIBUTE...")


def mission_success() -> None:
    print(f"DIRECT HIT!!!! {int(100 * random.random())} KILLED.")
    print("MISSION SUCCESSFUL.")


def death_with_chance(p_death: float) -> bool:
    """
    Takes a float between 0 and 1 and returns a boolean
    if the player has survived (based on random chance)

    Returns True if death, False if survived
    """
    return p_death > random.random()


def commence_non_kamikazi_attack() -> None:
    nmissions = int(input("HOW MANY MISSIONS HAVE YOU FLOWN? "))

    while nmissions >= 160:
        print("MISSIONS, NOT MILES...")
        print("150 MISSIONS IS HIGH EVEN FOR OLD-TIMERS")
        nmissions = int(input("NOW THEN, HOW MANY MISSIONS HAVE YOU FLOWN? "))

    if nmissions >= 100:
        print("THAT'S PUSHING THE ODDS!")

    if nmissions < 25:
        print("FRESH OUT OF TRAINING, EH?")

    print()
    return (
        mission_success() if nmissions >= 160 * random.random() else mission_failure()
    )


def mission_failure() -> None:
    weapons_choices = {
        "1": "GUNS",
        "2": "MISSILES",
        "3": "BOTH",
    }
    print(f"MISSED TARGET BY {int(2 + 30 * random.random())} MILES!")
    print("NOW YOU'RE REALLY IN FOR IT !!")
    print()
    enemy_weapons = _stdin_choice(
        prompt="DOES THE ENEMY HAVE GUNS(1), MISSILES(2), OR BOTH(3)? ",
        choices=weapons_choices.keys(),
    )

    # If there are no gunners (i.e. weapon choice 2) then
    # we say that the gunners have 0 accuracy for the purposes
    # of calculating probability of player death

    enemy_gunner_accuracy = 0.0
    if enemy_weapons != "2":
        # If the enemy has guns, how accurate are the gunners?
        enemy_gunner_accuracy = float(
            input("WHAT'S THE PERCENT HIT RATE OF ENEMY GUNNERS (10 TO 50)? ")
        )

        if enemy_gunner_accuracy < 10:
            print("YOU LIE, BUT YOU'LL PAY...")
            return player_death()

    missile_threat_weighting = 0 if enemy_weapons == "1" else 35

    death = death_with_chance(
        p_death=(enemy_gunner_accuracy + missile_threat_weighting) / 100
    )

    return player_survived() if not death else player_death()


def play_italy() -> None:
    targets_to_messages = {
        # 1 - ALBANIA, 2 - GREECE, 3 - NORTH AFRICA
        "1": "SHOULD BE EASY -- YOU'RE FLYING A NAZI-MADE PLANE.",
        "2": "BE CAREFUL!!!",
        "3": "YOU'RE GOING FOR THE OIL, EH?",
    }
    target = _stdin_choice(
        prompt="YOUR TARGET -- ALBANIA(1), GREECE(2), NORTH AFRICA(3)",
        choices=targets_to_messages.keys(),
    )

    print(targets_to_messages[target])
    return commence_non_kamikazi_attack()


def play_allies() -> None:
    aircraft_to_message = {
        "1": "YOU'VE GOT 2 TONS OF BOMBS FLYING FOR PLOESTI.",
        "2": "YOU'RE DUMPING THE A-BOMB ON HIROSHIMA.",
        "3": "YOU'RE CHASING THE BISMARK IN THE NORTH SEA.",
        "4": "YOU'RE BUSTING A GERMAN HEAVY WATER PLANT IN THE RUHR.",
    }
    aircraft = _stdin_choice(
        prompt="AIRCRAFT -- LIBERATOR(1), B-29(2), B-17(3), LANCASTER(4): ",
        choices=aircraft_to_message.keys(),
    )

    print(aircraft_to_message[aircraft])
    return commence_non_kamikazi_attack()


def play_japan() -> None:
    print("YOU'RE FLYING A KAMIKAZE MISSION OVER THE USS LEXINGTON.")
    first_mission = input("YOUR FIRST KAMIKAZE MISSION? (Y OR N): ")
    if first_mission.lower() == "n":
        return player_death()

    if random.random() > 0.65:
        return mission_success()
    return player_death()


def play_germany() -> None:
    targets_to_messages = {
        # 1 - RUSSIA, 2 - ENGLAND, 3 - FRANCE
        "1": "YOU'RE NEARING STALINGRAD.",
        "2": "NEARING LONDON.  BE CAREFUL, THEY'VE GOT RADAR.",
        "3": "NEARING VERSAILLES.  DUCK SOUP.  THEY'RE NEARLY DEFENSELESS.",
    }
    target = _stdin_choice(
        prompt="A NAZI, EH?  OH WELL.  ARE YOU GOING FOR RUSSIA(1),\nENGLAND(2), OR FRANCE(3)? ",
        choices=targets_to_messages.keys(),
    )

    print(targets_to_messages[target])

    return commence_non_kamikazi_attack()


def play_game() -> None:
    print("YOU ARE A PILOT IN A WORLD WAR II BOMBER.")
    sides = {"1": play_italy, "2": play_allies, "3": play_japan, "4": play_germany}
    side = _stdin_choice(
        prompt="WHAT SIDE -- ITALY(1), ALLIES(2), JAPAN(3), GERMANY(4): ",
        choices=sides.keys(),
    )
    return sides[side]()


if __name__ == "__main__":
    again = True
    while again:
        play_game()
        again = True if input("ANOTHER MISSION? (Y OR N): ").upper() == "Y" else False
